### ISSUE TITLE
Liberty support for configuring multiple SSL protocols

### ISSUE THREAD

**Issue by acdemyers** [OpenLiberty/open-liberty#19324]
## Description of the high level feature, including any external spec links:  
<br/><br/><br/>

Currently in the Liberty SSL configuration only one value can be added as the sslProtocol. There are values that include multiple protocols but most of these include all protocols.  This feature will allow users to create a narrow scope of protocols to support like TLSv1.2 and TLSv1.3.

The proposed new configuration will look something like:

<ssl id="defaultSSLConfig" 
             clientAuthenticationSupported="true" 
             sslProtocol="TLSv1.2,TLSv1.3"              
             keyStoreRef="defaultKeyStore">
</ssl>

Link to RFE: https://www.ibm.com/developerworks/rfe/execute?use_case=viewRfe&CR_ID=146872

##
#### Before proceeding to any items below (active development), this feature must be [prioritized on the backlog](https://github.ibm.com/orgs/websphere/projects/2), and have been socialized (UFO Review).  You may not deliver GM code until the feature Epic has obtained the "Design Approved" or "No Design Approved" label.  Follow the [Feature and UFO (WAD) Approval Process](https://github.ibm.com/websphere/WS-CD-Open/wiki/Feature-and-UFO-(WAD)-Approval-Process).
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the [UFO (Upcoming Feature Overview)](https://ibm.ent.box.com/file/318494237728) document and [FTS (Feature Test Summary)](https://github.ibm.com/websphere/WS-CD-Open/wiki/Feature-Review-(Feature-Test-Summary-Process)):
- UFO:  https://ibm.ent.box.com/file/878566993029
- FTS: #20670

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [ ] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [ ] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting WebSphere Liberty & tWAS](https://github.ibm.com/websphere/WS-CD-Open/wiki/Requesting-ID-Work-for-commercial-Liberty-and-tWAS)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.ibm.com/websphere/WS-CD-Open/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

## **Liberty Beta**
If your feature, or portions of it, are going to be included in a Liberty beta:  
Before Onboarding the beta
- [x] [Beta Fence](https://github.ibm.com/websphere/WS-CD-Open/wiki/Beta-Fencing) the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [ ] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [ ] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [ ] Implementation complete. (Epic owner / Feature owner)
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([websphere/fat-approvers](https://github.ibm.com/orgs/websphere/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

## **Other deliverbles**
- [ ] **WDT** - ([Leonard Theivendra](https://github.ibm.com/theivend)). WDT work complete or N/A.

**Comment by acdemyers** [comment]
Documentation issue https://github.com/OpenLiberty/docs/issues/5283

**Comment by acdemyers** [comment]
Serviceability questions:

UFO: Does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

The UFO mentions protocols are checked for valid protocol values when the configuration is process.  
If a list of protocols is configured and there is a protocol value not allowed in lists you users will see:

**CWPKI0832E: The SSL_TLS SSL/TLS protocol value that is provided cannot be used for configuring a list of SSL protocols values.
CWPKI0833E: The defaultSSLConfig SSL/TLS configuration has an error that prevents creation of the SSL/TLS configuration.** 

If a single protocol value is in the configuration and it's not allowed users will an error like this in the logs:

**CWPKI0831E: The SSLv2 SSL/TLS protocol cannot be used. Extended error is: SSLv2 SSLContext not available 
CWPKI0833E: The defaultSSLConfig SSL/TLS configuration has an error that prevents creation of the SSL/TLS configuration.** 


If the ssl-1.0 feature is used and the default SSL configuration there is error with the protocol an extra message is issued letting users know the process default SSL context will not be set:

CWPKI0834E: The defaultSSLConfig SSL/TLS configuration cannot be set as the process default SSL configuration due to an error.


Test and Demo: As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team). a) What problem paths were tested and demonstrated? b) Who did you demo to? c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

**It was demo-ed with the Security team.  The provided feedback on the messages.** 

SVT: SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered. a) Who conducted SVT tests for this feature? b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?

**Nichole Stuart did the SVT testing.  The serviceability is sufficient to avoid PMRs.**

Service: Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.

**WAS L2 Sec
WAS L3 Core Security**

Metrics: Does this feature add any new metrics or emit any new JSON events? If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

**n/a**

**Comment by chirp1** [comment]
Alaine provided info for ID to document in https://github.com/OpenLiberty/docs/issues/5283 Approving this epic.

**Comment by acdemyers** [comment]
STE loaded to https://ibm.ent.box.com/file/942780813025

**Comment by acdemyers** [comment]
Code included in PRs:
https://github.com/OpenLiberty/open-liberty/pull/19940
https://github.com/OpenLiberty/open-liberty/pull/20159
https://github.com/OpenLiberty/open-liberty/pull/20175
https://github.com/OpenLiberty/open-liberty/pull/20316

**Comment by donbourne** [comment]
I think there's a minor typo in this...
CWPKI0832E: The SSL_TLS SSL/TLS protocol value that is provided cannot be used for configuring a list of SSL protocols values.

should be
CWPKI0832E: The SSL_TLS SSL/TLS protocol value that is provided cannot be used for configuring a list of SSL protocol values.

Also, does the explanation for that message indicate why the value can't be used?

**Comment by acdemyers** [comment]
@donbourne 
I see,  I'll open an issue for the error message. Here is the full message group from the nlsprops file:
ssl.protocol.error.CWPKI0832E=CWPKI0832E: The {0} SSL/TLS protocol value that is provided cannot be used for configuring a list of SSL protocols values.
ssl.protocol.error.CWPKI0832E.explanation=The protocol value is not valid in a protocol list grouping.
ssl.protocol.error.CWPKI0832E.useraction=Ensure that the list of SSL/TLS protocols includes only values that are appropriate for a protocol list.

**Comment by donbourne** [comment]
where would someone go to find the valid list of protocols?  Can we say something about that in the explanation?

**Comment by acdemyers** [comment]
@donbourne 

Yes,  since I have to update the message.  I can update the message to do use a substitution for the list of valid protocols. I think I'm going to change the message to:

ssl.protocol.error.CWPKI0832E=CWPKI0832E: The {0} SSL/TLS protocol value that is provided cannot be used for configuring a list of SSL protocol values.  The following SSL/TLS protocols are valid: {1}
ssl.protocol.error.CWPKI0832E.explanation=The protocol value is not valid in a protocol list grouping.
ssl.protocol.error.CWPKI0832E.useraction=Ensure that the list of SSL/TLS protocols includes only values that are appropriate for a protocol list.
